### PR TITLE
[AIRFLOW-4971] Add Google Display & Video 360 integration

### PR DESCRIPTION
**Pull request by turbaszek**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4971

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: docs/howto/operator/gcp/display_video.rst [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+ .. Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+Google Display & Video 360 Operators
+=======================================
+Display & Video 360 has the end-to-end campaign management features you need. For more
```
Comment (mik-laj) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
`Google Display & Video 360 <https://marketingplatform.google.com/about/display-video-360/>`__ has the end-to-end campaign management features you need. 
```

File: docs/howto/operator/gcp/display_video.rst [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+information check `Google Display & Video 360 <https://marketingplatform.google.com/about/display-video-360/>`__
```
Comment (mik-laj) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion

```

File: docs/howto/operator/gcp/display_video.rst [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+.. exampleinclude:: ../../../../airflow/gcp/example_dags/example_display_video.py
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_google_display_video_createquery_report_operator]
+    :end-before: [END howto_google_display_video_createquery_report_operator]
+
+You can use :ref:`Jinja templating <jinja-templating>` with
+:template-fields:`airflow.gcp.operators.display_video.GoogleDisplayVideo360CreateReportOperator`
+parameters which allows you to dynamically determine values.
+The result is saved to :ref:`XCom <concepts:xcom>`, which allows it to be used by other operators.
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
The result is saved to :ref:`XCom <concepts:xcom>`, which allows the result to be used by other operators.
```

Comment (turbaszek) [reply docs/howto/operator/gcp/display_video.rst]: Thanks! 🐱

File: docs/howto/operator/gcp/display_video.rst [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+.. _howto/operator:GoogleDisplayVideo360DeleteReportOperator:
+
+Deleting a report
+^^^^^^^^^^^^^^^^^
+
+To delete Display&Video 360 report you can use
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
To delete Display&Video 360 report use
```

File: docs/howto/operator/gcp/display_video.rst [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+^^^^^^^^^^^^^^^^^^
+
+.. include:: _partials/prerequisite_tasks.rst
+
+.. _howto/operator:GoogleDisplayVideo360CreateReportOperator:
+
+Creating a report
+^^^^^^^^^^^^^^^^^
+
+To create Display&Video 360 report you can use
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
To create Display&Video 360 report use
```

File: docs/howto/operator/gcp/display_video.rst [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+:class:`~airflow.gcp.operators.display_video.GoogleDisplayVideo360CreateReportOperator`.
+
+.. exampleinclude:: ../../../../airflow/gcp/example_dags/example_display_video.py
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_google_display_video_createquery_report_operator]
+    :end-before: [END howto_google_display_video_createquery_report_operator]
+
+You can use :ref:`Jinja templating <jinja-templating>` with
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
Use :ref:`Jinja templating <jinja-templating>` with
```

File: docs/howto/operator/gcp/display_video.rst [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+:template-fields:`airflow.gcp.operators.display_video.GoogleDisplayVideo360CreateReportOperator`
+parameters which allows you to dynamically determine values.
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
parameters which allow you to dynamically determine values.
```

File: docs/howto/operator/gcp/display_video.rst [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+You can use :ref:`Jinja templating <jinja-templating>` with
+:template-fields:`airflow.gcp.sensors.display_video.GoogleDisplayVideo360ReportSensor`
+parameters which allows you to dynamically determine values.
+
+.. _howto/operator:GoogleDisplayVideo360DownloadReportOperator:
+
+Downloading a report
+^^^^^^^^^^^^^^^^^^^^
+
+To create download a report to GCS bucket you can use
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
To create download a report to GCS bucket you can use
``` Not sure what you meant. Was it "create" or "download" in the end?

Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
To create download a report to GCS bucket use
```

File: docs/howto/operator/gcp/display_video.rst [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+.. exampleinclude:: ../../../../airflow/gcp/example_dags/example_display_video.py
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_google_display_video_getquery_report_operator]
+    :end-before: [END howto_google_display_video_getquery_report_operator]
+
+You can use :ref:`Jinja templating <jinja-templating>` with
+:template-fields:`airflow.gcp.operators.display_video.GoogleDisplayVideo360DownloadReportOperator`
+parameters which allows you to dynamically determine values.
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
parameters which allow you to dynamically determine values.
```

File: docs/howto/operator/gcp/display_video.rst [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+To create download a report to GCS bucket you can use
+:class:`~airflow.gcp.operators.display_video.GoogleDisplayVideo360DownloadReportOperator`.
+
+.. exampleinclude:: ../../../../airflow/gcp/example_dags/example_display_video.py
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_google_display_video_getquery_report_operator]
+    :end-before: [END howto_google_display_video_getquery_report_operator]
+
+You can use :ref:`Jinja templating <jinja-templating>` with
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
Use :ref:`Jinja templating <jinja-templating>` with
```

File: docs/howto/operator/gcp/display_video.rst [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+:template-fields:`airflow.gcp.operators.display_video.GoogleDisplayVideo360DownloadReportOperator`
+parameters which allows you to dynamically determine values.
+
+
+.. _howto/operator:GoogleDisplayVideo360RunReportOperator:
+
+Running a report
+^^^^^^^^^^^^^^^^
+
+To run Display&Video 360 report you can use
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
To run Display&Video 360 report use
```

File: docs/howto/operator/gcp/display_video.rst [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+:class:`~airflow.gcp.operators.display_video.GoogleDisplayVideo360RunReportOperator`.
+
+.. exampleinclude:: ../../../../airflow/gcp/example_dags/example_display_video.py
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_google_display_video_runquery_report_operator]
+    :end-before: [END howto_google_display_video_runquery_report_operator]
+
+You can use :ref:`Jinja templating <jinja-templating>` with
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
Use :ref:`Jinja templating <jinja-templating>` with
```

File: docs/howto/operator/gcp/display_video.rst [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+:template-fields:`airflow.gcp.operators.display_video.GoogleDisplayVideo360RunReportOperator`
+parameters which allows you to dynamically determine values.
```
Comment (mschickensoup) [inline docs/howto/operator/gcp/display_video.rst]:
```suggestion
parameters which allow you to dynamically determine values.
```

Comment (turbaszek) [reply docs/howto/operator/gcp/display_video.rst]: Thanks @mschickensoup!

Comment (mschickensoup) [reply docs/howto/operator/gcp/display_video.rst]: YW!

File: airflow/gcp/operators/display_video.py [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+            google_cloud_storage_conn_id=self.gcp_conn_id, delegate_to=self.delegate_to
+        )
+
+        self.log.info("Starting downloading report %s", self.report_id)
+        resource = hook.get_query(query_id=self.report_id)
+        file_url = resource["metadata"]["googleCloudStoragePathForLatestReport"]
+
+        # Download the report
+        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+            with urllib.request.urlopen(file_url) as response:
```
Comment (potiuk) [inline airflow/gcp/operators/display_video.py]: Are you sure this will not load the whole file into memory ? It likely does but I am not 100 % sure.  I guess these reports can be quite big so it might be worth doing streaming here.

Comment (turbaszek) [reply airflow/gcp/operators/display_video.py]: It seems to be a recommended way of doing things:
https://docs.python.org/3/howto/urllib2.html#fetching-urls

If I fully understand `urlopen` does not load the file into memory and `shutil.copyfileobj`:
>Copy the contents of the file-like object fsrc to the file-like object fdst. ... by default the data is read in chunks to avoid uncontrolled memory consumption.

https://docs.python.org/3/library/shutil.html#shutil.copyfileobj

Comment (potiuk) [reply airflow/gcp/operators/display_video.py]: Good then :)

File: airflow/gcp/operators/display_video.py [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+    :param api_version: The version of the api that will be requested for example 'v3'.
+    :type api_version: str
+    :param gcp_conn_id: The connection ID to use when fetching connection info.
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any. For this to work, the service accountmaking the
+        request must have  domain-wide delegation enabled.
+    :type delegate_to: str
+    """
+
+    template_fields = ("body",)
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: We should support json as a template extension.
```suggestion
    template_fields = ("body",)
    template_ext = (".json",)
```

File: airflow/gcp/operators/display_video.py [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+        )
+        self.log.info("Creating Display & Video 360 report.")
+        response = hook.create_query(query=self.body)
+        report_id = response["queryId"]
+        self.xcom_push(context, key="report_id", value=report_id)
+        self.log.info("Created report with ID: %s", report_id)
+        return response
+
+
+class GoogleDisplayVideo360DeleteReportOperator(BaseOperator):
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: Could we also add the capability to delete reports by title?

File: airflow/gcp/operators/display_video.py [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+
+    template_fields = ("report_id", "bucket_name", "report_name")
+
+    @apply_defaults
+    def __init__(
+        self,
+        report_id: str,
+        bucket_name: str,
+        report_name: str,
+        gzip: bool = False,
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: Can we default compression to True as these reports can get quite large?
```suggestion
        gzip: bool = False,
        gzip: bool = True,
```

File: airflow/gcp/operators/display_video.py [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+    """
+
+    template_fields = ("report_id", "bucket_name", "report_name")
+
+    @apply_defaults
+    def __init__(
+        self,
+        report_id: str,
+        bucket_name: str,
+        report_name: str,
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: Can we make report_name optional and default it to the report name from DV360?

File: airflow/gcp/operators/display_video.py [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+        gzip: bool = False,
+        chunk_size: int = 5 * 1024 * 1024,
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: I'd also up the default chunk size to 10MB to better support larger file downloads.

File: airflow/gcp/operators/display_video.py [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+        self.gzip = gzip
+        self.bucket_name = self._set_bucket_name(bucket_name)
+        self.report_name = self._set_report_name(report_name)
+        self.api_version = api_version
+        self.gcp_conn_id = gcp_conn_id
+        self.delegate_to = delegate_to
+
+    @staticmethod
+    def _set_report_name(name: str) -> str:
+        return name if name.endswith(".csv") else name + ".csv"
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: In the case of compressed reports, should also add the "*.gz" extension.

File: airflow/gcp/operators/display_video.py [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+            delegate_to=self.delegate_to,
+            api_version=self.api_version,
+        )
+        gcs_hook = GoogleCloudStorageHook(
+            google_cloud_storage_conn_id=self.gcp_conn_id, delegate_to=self.delegate_to
+        )
+
+        self.log.info("Starting downloading report %s", self.report_id)
+        resource = hook.get_query(query_id=self.report_id)
+        file_url = resource["metadata"]["googleCloudStoragePathForLatestReport"]
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: You may want to check that the report is not currently running `resource["metadata"]["running"]` before attempting to download the report.

File: airflow/gcp/operators/display_video.py [hunk 22]
```diff
@@ -0,0 +1,301 @@
...
+                filename=temp_file.name,
+                mime_type="text/csv",
+            )
+        self.log.info(
+            "Report %s was saved in bucket %s as %s.",
+            self.report_id,
+            self.bucket_name,
+            self.report_name,
+        )
+
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: We should push report_name as an XCom, especially if we make report_name optional.

File: airflow/gcp/operators/display_video.py [hunk 23]
```diff
@@ -0,0 +1,301 @@
...
+        self.log.info("Starting downloading report %s", self.report_id)
+        resource = hook.get_query(query_id=self.report_id)
+        file_url = resource["metadata"]["googleCloudStoragePathForLatestReport"]
+
+        # Download the report
+        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+            with urllib.request.urlopen(file_url) as response:
+                shutil.copyfileobj(response, temp_file)
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: We should push the chunk_size as the length so that we can tune performance.

File: airflow/gcp/operators/display_video.py [hunk 24]
```diff
@@ -0,0 +1,301 @@
...
+    :type delegate_to: str
+    """
+
+    template_fields = ("report_id", "params")
+
+    @apply_defaults
+    def __init__(
+        self,
+        report_id: str,
+        params: Dict[str, Any],
```
Comment (efolgar) [inline airflow/gcp/operators/display_video.py]: Could we rename this (e.g. "body") as params is typically reserved for Jinja templating?

File: airflow/gcp/sensors/display_video.py [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+    :param gcp_conn_id: The connection ID to use when fetching connection info.
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any. For this to work, the service accountmaking the
+        request must have  domain-wide delegation enabled.
+    :type delegate_to: str
+    """
+
+    template_fields = ("report_id",)
+
+    def __init__(
```
Comment (efolgar) [inline airflow/gcp/sensors/display_video.py]: Can we default the mode to "reschedule" and set the default poke interval to 5 minutes to better support larger report sizes?

**Review by efolgar — CHANGES_REQUESTED** [review]
Reviewed the hook, operators, and sensor for DV360.

File: airflow/gcp/operators/display_video.py (hunk 18 above)
Comment (turbaszek) [reply airflow/gcp/operators/display_video.py]: What should be treated as default name in this case? Should it be `queryId`?

Comment (efolgar) [reply airflow/gcp/operators/display_video.py]: The report name can be parsed from the file url.  Something like:

`file_url = resource["metadata"]["googleCloudStoragePathForLatestReport"]`
`report_name = urlparse(file_url).path.split('/')[2]`

**Comment by mik-laj** [comment]
@nuclearpinguin Can you add this operators to `operators-and-hooks-ref.rst` file?

File: tests/providers/google/marketing_platform/hooks/test_campaign_manager.py [hunk 26]
```diff
@@ -40,8 +40,7 @@ def setUp(self):
         "campaign_manager.GoogleCampaignManagerHook._authorize"
     )
     @mock.patch(
-        "airflow.providers.google.marketing_platform.hooks."
-        "campaign_manager.build"
+        "airflow.providers.google.marketing_platform.hooks." "campaign_manager.build"
```
Comment (mik-laj) [inline tests/providers/google/marketing_platform/hooks/test_campaign_manager.py]: Can you revert these changes because they make it difficult for cherry-pick these operators to another project/package/release/etc.?

File: tests/providers/google/marketing_platform/operators/test_campaign_manager.py [hunk 27]
```diff
@@ -56,8 +56,7 @@ def test_execute(self, mock_base_op, hook_mock):
 
 class TestGoogleCampaignManagerGetReportOperator(TestCase):
     @mock.patch(
-        "airflow.providers.google.marketing_platform.operators."
-        "campaign_manager.http"
+        "airflow.providers.google.marketing_platform.operators." "campaign_manager.http"
```
Comment (mik-laj) [inline tests/providers/google/marketing_platform/operators/test_campaign_manager.py]: Can you revert these changes because they make it difficult for cherry-pick these operators to another project/package/release/etc.?

File: tests/providers/google/marketing_platform/hooks/test_campaign_manager.py (hunk 26 above)
Comment (turbaszek) [reply tests/providers/google/marketing_platform/hooks/test_campaign_manager.py]: Done

File: tests/providers/google/marketing_platform/operators/test_campaign_manager.py (hunk 27 above)
Comment (turbaszek) [reply tests/providers/google/marketing_platform/operators/test_campaign_manager.py]: Done
